energy_calc/get_pulley_torque_fun.m

```diff
@@ -23,45 +32,135 @@
+        new_point = last_point + last_anchor_to_pulley/last_free_len * alpha_dists(i-1);
+        anchor_to_pulley = new_point-anchor_ps(i,:);
+        
+        free_len = norm(anchor_to_pulley);
+        prev_len = alpha_dists(i-1) + last_free_len;
+        x_diff = free_len - prev_len;
         
-        if draw %save points
-            pulley_ps(i-1,:) = alpha;
-        end
-
         %new radius is distance from pull line to origin
-        cross2d = @(a,b) cross(horzcat(a,0),horzcat(b,0));
-        r = norm(cross2d(alpha-anchor_ps(i,:),-anchor_ps(i,:)))/norm(alpha-anchor_ps(i,:));
+        lever_arm = abs(cross2d(anchor_to_pulley,-anchor_ps(i,:)))/free_len;
 
-        f_spring = last_f_spring + k_spring*x_diff;
-        torques(i) = f_spring * r;
+        f_spring = max(0, last_f_spring + k_spring*x_diff); %force can't go below 0
+        torques(i) = f_spring * lever_arm;
         
+        % Store values for constraints
+        max_f_spring = max(max_f_spring, f_spring);
+        min_f_spring = min(min_f_spring, f_spring);
+        r = norm(new_point);
+        max_pulley_r = max(max_pulley_r, r);
+        min_pulley_lever = min(min_pulley_lever, lever_arm);
+        min_free_len = min(min_free_len, free_len);
+        min_angle_to_origin = min(min_angle_to_origin,cross2d(-anchor_ps(i,:),anchor_to_pulley));        
+        
+        if new_point(1)>0 && new_point(2)<0 %bottom right quadrant? forbidden
+            worst_bottom_right_quadrant = max(worst_bottom_right_quadrant, new_point(1) - new_point(2));
+        end
+        
+        % Store values for drawing
+        if draw
+            pulley_ps(i-1,:) = new_point;
+            f_springs(i-1) = f_spring;
+        end
+        
+        % Store calculated values for next loop
         last_f_spring = f_spring;
+        last_point = new_point;
+        last_anchor_to_pulley = anchor_to_pulley;
+        last_free_len = free_len;
     end
-    
+     
     if draw
         figure(100);
         cla reset;
         axis equal
         hold on
         plot(anchor_ps(:,1),anchor_ps(:,2),'b');
-        plot([pulley_ps(:,1); 0; pulley_ps(1,1)],[pulley_ps(:,2); 0; pulley_ps(1,2)],'r.'); 
+        connect_origin = false;
+        if connect_origin
+            plot([pulley_ps(:,1); 0; pulley_ps(1,1)],[pulley_ps(:,2); 0; pulley_ps(1,2)],'r.-'); 
+        else
+            plot(pulley_ps(:,1),pulley_ps(:,2),'r.-'); 
+        end
         plot([0 0 -anchor_offset],[0 -anchor_dist -anchor_dist], 'k', 'LineWidth',2)
         plot(px_1,0,'k.');
         for i = [1 floor(size(pulley_ps,1)/2) size(pulley_ps,1)]
             plot([pulley_ps(i,1) anchor_ps(i,1)],[pulley_ps(i,2) anchor_ps(i,2)],'k--')
         end
         pad = diff(ylim)*0.1;
         ylim(ylim + [-pad pad]);
+        xlabel('x (m)');
+        ylabel('y (m)');
+        
+        figure(90);
+        cla reset;
+        axis equal
+        
+        plot(thetas(1:end-1), imp(f_springs,'N')); %TODO: why diff len?
```

Comment:
although this should be worked out, I think somewhere there is a mislabeling of 1:end and 2:end and 1:end-1 etc.